ls1intum/Artemis
# `General`: Fix active students charts with limited scope

### Checklist
#### General


#### Server

### Motivation and Context


This PR is concerning Issue #5101: Currently, the scope limitation of the active students charts does not work as intended. If a course has a start date and the difference between the course start date and the current end date of the chart (either the official end date of the course or the current date) was not exactly a multiple of a week, the chart did not show the correct information.
Furthermore, the server test regarding the scope limitation server code is currently flaky.

### Description

I adapted the server code. Now, the difference between start and end dates is computed based on normalized timestamps.
Additionally, I modified the server test so that it passes again.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with student activity

1. Log in to Artemis
2. Edit the course
3. Set a course start date that is  e.g. 20 days ago.
4. Safe the course and check whether the chart on course detail shows the correct information
5. Check that the chart in the course management tile shows correct information as well
6. Edit the course again. Choose the start date so that there is a difference of > 4 weeks between it and the current date. **Still, it should not be the current time exactly X weeks ago**
7. Check the charts again that they still work
8. Edit the course again. Set a course end date that lies in the past. Play around with the start date so that the time difference is not exactly a multiple of a week. Make sure that both charts still work and show correct data.
9. Also try a combination that leads to a course lifetime > 1 year and verify that the charts handle this correctly.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|


### Screenshots


The scope limitation should work again:
![image](https://user-images.githubusercontent.com/80622272/168620978-382090da-0aaa-4d2c-8577-5bd0d7bcfc5c.png)
